### ISSUE TITLE
Rethink the Checkpoint Loading function

### ISSUE THREAD

**Issue by dlwh** [stanford-crfm/levanter#481]
As part of my goal to rethink the TrainerState abstraction #420, I'm trying to figure out how best to encapsulate checkpoint loading.

Checkpoint saving is easy: you have a PyTree of data and we pass it to serialize_leaves_tensorstore. The only subtlety is that we only serialize the `is_trainable` parts of the model (so that we only serialize the lora adapter weights and not the whole model.)

Deserializing is a bit more complex, because it's coupled to initialization. In order to deserialize something, we need to know the shape, so we have to at least do an `eval_shape` on the initialization code. So at a minimum we need something like this:

```
def initialize(): ...

if checkpoint_exists:
   shape = filter_eval_shape(initialize)
   state = load_checkpoint(...)
else:
    state = jit(initialize)()
```

So far so good, but our initialization story is more complicated when we think about model initialization.
We can initialize a model 3 ways: from scratch, from a levanter checkpoint, or from a HF checkpoint.
Unfortunately the last two are very different codepaths and probably will be for some time. (The HF checkpoint codepaths support manipulating the array shapes in the checkpoint to Levanter's model structure, but we don't do that currently for Levanter's internal checkpoints.) So now the flow looks like:

```
def initialize_model(): ...
def initialize_state(model): ...

if state_checkpoint_exists:
    model_shape = eval_shape(initialize_model)
    state_shape = eval_shape(initialize_state, model_shape)
    state = load_checkpoint(...)
else:
    if lev_model_checkpoint_exists:
        model_shape = eval_shape(initialize_model)
        model = load_checkpoint(...)
   elif hf_checkpoint:
        model_shape = eval_shape(initialize_model)
        model = load_hf_checkpoint(...)
   else:
        model = jit(initialize_model)()
   
   state = jit(initialize_state)(model) 
```

Not terrible, but 2 things:

1. atm we keep the hf_checkpoint stuff away from the main trainer class, so we can't directly use this code.
2. We also need to support "partial checkpoint" initialization for stuff like llama, where some buffers are rematerialized from scratch when we load the model.

(2) makes the flow look like this:

```
def initialize_model(): ...
def initialize_state(model): ...


# if run in jit, this won't waste non-symbolic work as long at the compiler does its job of DCE
def full_init(init_fn, partial_result):
    partial_result = filter_out_any_abstract_shapes(partial_result)
    return eqx.combine(partial_result, init_fn()):

if state_checkpoint_exists:
    model_shape = eval_shape(initialize_model)
    state_shape = eval_shape(initialize_state, model_shape)
    partial_state = load_checkpoint(...)

    state = jit(full_init)(lambda : initialize_state(initialize_model()), partial_state)
else:
    if lev_model_checkpoint_exists:
        model_shape = eval_shape(initialize_model)
        partial_model = load_checkpoint(...)
   elif hf_checkpoint:
        model_shape = eval_shape(initialize_model)
        partial_model = load_hf_checkpoint(...)
   else:
        partial_model = eval_shape(initialize_model)
   
   model = jit(full_init)(initialize_model, partial_model)
   state = jit(initialize_state)(model) 
```

Ok, we're getting there.

I was trying to make a method like

```
def load_or_initialize(init_fn, path):
   shape = eval_shape(init_fn)
   if exists(path):
       partial_result = load_checkpoint(shape)
   else:
       partial_result = shape
   
   return jit(full_init)(init_fn, partial_result)
```

But that doesn't end up composing b/c we can't do i/o inside jit, so we can't do something pretty like:

```
# DOESN'T WORK

def maybe_init_state_from_model_ckpt(init_model):
     model = load_or_initialize(init_model_maybe_from_hf, path)
     state = initialize_state(model)
     return state

state = load_or_initialize(maybe_int_state_from_model_ckpt, ...)
```

Sad :(

**Comment by dlwh** [comment]
I think what we do is make `load_or_initialize(...)`, push the initial_state out of Trainer (which we want to do anyway), if you need this fancy two stage init, you just have to check for checkpoint existence yourself.

**Comment by dlwh** [comment]
fixed in #496
